### PR TITLE
Prevent Unrelated Casts on Child Choice Node

### PR DESCRIPTION
**Pull request by Matejkob**
Resolves part of #2092

### CODE REVIEW

**Review by ahoppen — COMMENTED** [review]
Hmm, this does significantly reduce the readability in many cases. I didn’t consider that you couldn’t easily chain `if case`…

After reading this, I think I changed my mind and we should offer non-deprecated casting methods to any option that the `SyntaxChildChoices` can have. And if one of that option is a base kind like `DeclSyntax`, we should offer a casting method to any `DeclSyntaxProtocol`.

**Comment by Matejkob** [comment]
> Hmm, this does significantly reduce the readability in many cases. I didn’t consider that you couldn’t easily chain if case…

I completely agree. That was my main concern with value binding in the first place. I haven't yet found the time to go through this pitch: [https://forums.swift.org/t/pitch-is-case-expressions/64185](https://forums.swift.org/t/pitch-is-case-expressions/64185), so I'm not sure why it hasn't gone through yet. The issue definitely should be addressed—it's a day-to-day problem for many folks.

> After reading this, I think I changed my mind and we should offer non-deprecated casting methods to any option that the SyntaxChildChoices can have. And if one of that option is a base kind like DeclSyntax, we should offer a casting method to any DeclSyntaxProtocol.

Sounds good to me. I'll try cook something

**Comment by ahoppen** [comment]
> so I'm not sure why it hasn't gone through yet. The issue definitely should be addressed—it's a day-to-day problem for many folks.

I haven’t read the pitch completely either but I would imagine it’s just a matter of someone writing the official proposal and creating an implementation for it.

**Comment by Matejkob** [comment]
Base on last John's post https://forums.swift.org/t/pitch-is-case-expressions/64185/69 the whole pitch is abandoned 😢

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxNodeCasting.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import SwiftSyntax
+import SwiftSyntaxBuilder
+import SyntaxSupport
+
+@MemberBlockItemListBuilder
+func choiceNodeCastingMethods(for syntaxNodeKind: SyntaxNodeKind) -> MemberBlockItemListSyntax {
+  if syntaxNodeKind.isBase {
+    DeclSyntax(
+      """
+      public func `is`<S: \(syntaxNodeKind.protocolType)>(_ syntaxType: S.Type) -> Bool {
```
Comment (Matejkob) [inline CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxNodeCasting.swift:22]: How about adding to those methods documentation in the same manner as to others e.g. https://github.com/apple/swift-syntax/blob/00dc4d966c25459a95cc65d4cab11013616a2d63/Sources/SwiftSyntax/SyntaxProtocol.swift#L133-L143 ?

File: Release Notes/510.md [hunk 2]
```diff
@@ -49,6 +49,11 @@
   - Description: `is`, `as`, and `cast` methods on base node protocols with base-type conversions are marked as deprecated. The deprecated methods emit a warning that informs the developer that the cast will always succeed and should be done using the base node's initializer.
   - Issue: https://github.com/apple/swift-syntax/issues/2092
   - Pull Request: https://github.com/apple/swift-syntax/pull/2108
+  
+- Child Choice Node Casts
+  - Description: TODO: Add!
```
Comment (Matejkob) [inline Release Notes/510.md:54]: Gonna add at the end

**Comment by Matejkob** [comment]
@ahoppen how about now? Usage is cleaner, but there is plenty of code more on the implementation part :/

File: CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxNodeCasting.swift (hunk 1 above)
Comment (kimdv) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxNodeCasting.swift:22]: Sounds good for me

Comment (ahoppen) [reply CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/SyntaxNodeCasting.swift:22]: Yes, those doc comments would be useful.

**Review by ahoppen — COMMENTED** [review]
Looks good to me in general

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by kimdv** [comment]
@swift-ci please test windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**